plugins/initialization-control/src/main/java/org/qbicc/plugin/initializationcontrol/FeaturePatcher.java

```diff
@@ -42,44 +45,139 @@ public boolean isRuntimeInitializedClass(String internalName) {
     }
 
     public boolean isReflectiveField(String className, String fieldName) {
+        ClassInfo ci = reflectiveClasses.get(className);
+        if (ci != null && ci.fields) {
+            return true;
+        }
         Set<String> fields = reflectiveFields.get(className);
         return fields != null && fields.contains(fieldName);
     }
 
-    public void addReflectiveMethod(String className, String methodName, String descriptor) {
-        String encodedMethod = methodName+":"+descriptor;
-        reflectiveMethods.computeIfAbsent(className, k -> ConcurrentHashMap.newKeySet()).add(encodedMethod);
+    public void addReflectiveMethod(String className, String methodName, String[] parameterTypes) {
+        reflectiveMethods.computeIfAbsent(className, k -> ConcurrentHashMap.newKeySet()).add(methodName+":"+encodeArguments(parameterTypes));
     }
 
     public boolean hasReflectiveMethods(String className) {
-        return reflectiveMethods.containsKey(className);
+        ClassInfo ci = reflectiveClasses.get(className);
+        return (ci != null && ci.methods) || reflectiveMethods.containsKey(className);
     }
 
     public boolean isReflectiveMethod(String className, String methodName, MethodDescriptor descriptor) {
+        ClassInfo ci = reflectiveClasses.get(className);
+        if (ci != null && ci.methods) {
+            return true;
+        }
         Set<String> encodedMethods = reflectiveMethods.get(className);
-        return encodedMethods != null && encodedMethods.contains(methodName+":"+descriptor);
+        if (encodedMethods != null) {
+            for (String candidate : encodedMethods) {
+                String[] split = candidate.split(":");
+                if (methodName.equals(split[0]) && matchesArguments(candidate, descriptor.getParameterTypes())) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private String encodeArguments(String[] args) {
+        if (args == null || args.length == 0) {
+            return "";
+        }
+        if (args.length == 1 && args.equals("*")) {
+            return "*"; // wildcard -- matches all args
+        }
+        String ans = toDescriptorString(args[0]);
+        for (int i = 1; i<args.length; i++) {
+            ans = ans + "," + toDescriptorString(args[i]);
+        }
+        return ans;
+    }
+
+    private String toDescriptorString(String t) {
+        return switch (t) {
+            case "boolean" -> "Z";
+            case "byte" -> "B";
+            case "short" -> "S";
+            case "char" -> "C";
+            case "int" -> "I";
+            case "float" -> "F";
+            case "long" -> "J";
+            case "double" -> "D";
+            // TODO: friendlier processing of array types -- for now we just force the user to provide a descriptor if they want an array
+            default -> t.startsWith("[") ? t : "J"+t.replace('.','/')+";";
```

Comment:
I guess this was meant to be: 
```suggestion
            default -> t.startsWith("[") ? t : "L"+t.replace('.','/')+";";
```
?

Reply:
🤦 yeah, it was meant to be a descriptor.